torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
   return ncclMemAllocator;
 }
 
+at::Tensor ProcessGroupNCCL::allocateTensor(
+    long size,
+    at::TensorOptions options) {
+  // Some checks
+  TORCH_CHECK_VALUE(options.has_device(), "Tensor options must include device");
+  auto device = options.device();
+  TORCH_CHECK_VALUE(
+      device.is_cuda(),
+      "NCCL tensor allocator expects cuda type but got " + c10::str(device))
+
+  at::cuda::OptionalCUDAGuard gpuGuard(device);
+
+  // Create memory pool
+  if (!memPool_) {
+    // Needs a CUDAAllocator
+    auto allocator =
+        reinterpret_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator*>(
+            getMemAllocator().get());
+    // Pool is created
+    memPool_ = std::make_unique<c10::cuda::MemPool>(allocator);
+    // Also need to ncclCommRegister it so that NCCL recognizes it
+    registerMemPool(memPool_.get());
```

Comment:
This call should happen *after* the tensor is allocated. Inside `registerMemPool`, we call snapshot, which returns all the blocks held by the pool. The pointers of those blocks are then registered with NCCL via ncclCommRegister. So right now, this is registering nothing.